File: CommonBusiness/src/main/java/com/company/commonbusiness/http/RetrofitFactory.java
package com.company.commonbusiness.http;

import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @author 李昭鸿
 * @desc: Retrofit配置类
 * @date Created on 2017/7/26 16:31
 */

public class RetrofitFactory {

    private static final String BASE_URL = "https://precustomerservicebackend.crlandpm.com.cn";
    private static final int DEFAULT_TIMEOUT = 20;

    private boolean isTrustAll = true;
    private static RetrofitFactory instance;
    private final Retrofit retrofit;

    private RetrofitFactory() {
        retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(setOkHttpClient())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
    }

    public static RetrofitFactory getInstance() {
        if (null == instance) {
            synchronized (RetrofitFactory.class) {
                if (null == instance) {
                    instance = new RetrofitFactory();
                }
            }
        }

        return instance;
    }

    /**
     * 配置OkHttp
     *
     * @return 自定义的OkHttp
     */
    public OkHttpClient setOkHttpClient() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        builder.addInterceptor(interceptor)
                .connectTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .readTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .retryOnConnectionFailure(true);

        // 设置https信任所有的证书
        if (isTrustAll) {
            X509TrustManager trustManager = new X509TrustManager() {
                @Override
                public void checkClientTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {

                }

                @Override
                public void checkServerTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {

                }

                @Override
                public X509Certificate[] getAcceptedIssuers() {
                    return new java.security.cert.X509Certificate[]{};
                }
            };

            try {
                final SSLContext sslContext = SSLContext.getInstance("TLS");
                sslContext.init(null, new TrustManager[] { trustManager }, null);
                SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
                builder.sslSocketFactory(sslSocketFactory, trustManager);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        return builder.build();
    }

    /**
     * 返回接口Service
     *
     * @param service 接口服务
     * @param <T> 泛型
     * @return
     */
    public <T> T initService(final Class<T> service) {
        if (service == null) {
            throw new RuntimeException("Api service is null!");
        }
        return retrofit.create(service);
    }

}

File: CommonBusiness/src/main/java/com/company/commonbusiness/http/RetrofitFactory.kt
package com.company.commonbusiness.http

import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit

import javax.net.ssl.SSLContext
import javax.net.ssl.SSLSocketFactory
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

/**
 * @author 李昭鸿
 * @desc: Retrofit配置类
 * @date Created on 2017/7/26 16:31
 */

class RetrofitFactory private constructor() {

    private val isTrustAll = true
    private val builder: Retrofit.Builder

    init {
        builder = Retrofit.Builder()
            .client(setOkHttpClient())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(GsonConverterFactory.create())
    }

    /**
     * 配置OkHttp
     *
     * @return 自定义的OkHttp
     */
    fun setOkHttpClient(): OkHttpClient {
        val interceptor = HttpLoggingInterceptor()
        interceptor.level = HttpLoggingInterceptor.Level.BODY

        val builder = OkHttpClient.Builder()
        builder.addInterceptor(interceptor)
            .connectTimeout(DEFAULT_TIMEOUT.toLong(), TimeUnit.SECONDS)
            .readTimeout(DEFAULT_TIMEOUT.toLong(), TimeUnit.SECONDS)
            .retryOnConnectionFailure(true)

        // 设置https信任所有的证书
        if (isTrustAll) {
            val trustManager = object : X509TrustManager {
                @Throws(CertificateException::class)
                override fun checkClientTrusted(
                    x509Certificates: Array<X509Certificate>,
                    s: String
                ) {

                }

                @Throws(CertificateException::class)
                override fun checkServerTrusted(
                    x509Certificates: Array<X509Certificate>,
                    s: String
                ) {

                }

                override fun getAcceptedIssuers(): Array<X509Certificate> {
                    return arrayOf()
                }
            }

            try {
                val sslContext = SSLContext.getInstance("TLS")
                sslContext.init(null, arrayOf<TrustManager>(trustManager), null)
                val sslSocketFactory = sslContext.socketFactory
                builder.sslSocketFactory(sslSocketFactory, trustManager)
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }

        return builder.build()
    }

    /**
     * 返回接口Service
     *
     * @param service 接口服务
     * @param <T> 泛型
     * @return
    </T> */
    fun <T> initService(service: Class<T>?, hostUrl: String): T {
        if (service == null) {
            throw RuntimeException("Api service is null!")
        }

        val retrofit = builder.baseUrl(hostUrl).build()

        return retrofit.create(service)
    }

    companion object {

        private val DEFAULT_TIMEOUT = 20
        private var instance: RetrofitFactory? = null

        fun getInstance(): RetrofitFactory {
            if (null == instance) {
                synchronized(RetrofitFactory::class.java) {
                    if (null == instance) {
                        instance = RetrofitFactory()
                    }
                }
            }

            return instance!!
        }
    }

}
